homeassistant/components/osoenergy/water_heater.py

```diff
@@ -46,14 +58,116 @@ async def async_setup_entry(
         return
     async_add_entities((OSOEnergyWaterHeater(osoenergy, dev) for dev in devices), True)
 
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_GET_PROFILE,
+        {},
+        OSOEnergyWaterHeater.async_get_profile.__name__,
+        supports_response=SupportsResponse.ONLY,
+    )
+
+    service_set_profile_schema = cv.make_entity_service_schema(
+        {
+            vol.Optional(f"hour_{hour:02d}"): vol.All(
+                vol.Coerce(int), vol.Range(min=10, max=75)
+            )
+            for hour in range(24)
+        }
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_SET_PROFILE,
+        service_set_profile_schema,
+        OSOEnergyWaterHeater.async_set_profile.__name__,
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_SET_V40MIN,
+        {
+            vol.Required(ATTR_V40MIN): vol.All(
+                vol.Coerce(float), vol.Range(min=200, max=550)
+            ),
+        },
+        OSOEnergyWaterHeater.async_set_v40_min.__name__,
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_TURN_OFF,
+        {vol.Required(ATTR_UNTIL_TEMP_LIMIT): vol.All(cv.boolean)},
+        OSOEnergyWaterHeater.async_oso_turn_off.__name__,
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_TURN_ON,
+        {vol.Required(ATTR_UNTIL_TEMP_LIMIT): vol.All(cv.boolean)},
+        OSOEnergyWaterHeater.async_oso_turn_on.__name__,
+    )
+
+
+def _get_utc_hour(local_hour: int) -> dt.datetime:
+    """Convert the requested local hour to a utc hour for the day.
+
+    Args:
+        local_hour: the local hour (0-23) for the current day to be converted.
+
+    Returns:
+        Datetime representation for the requested hour in utc time for the day.
+
+    """
+    now = dt_util.now()
+    local_time = now.replace(hour=local_hour, minute=0, second=0, microsecond=0)
+    return dt_util.as_utc(local_time)
+
+
+def _get_local_hour(utc_hour: int) -> dt.datetime:
+    """Convert the requested utc hour to a local hour for the day.
+
+    Args:
+        utc_hour: the utc hour (0-23) for the current day to be converted.
+
+    Returns:
+        Datetime representation for the requested hour in local time for the day.
+
+    """
+    now = dt_util.utcnow()
+    now_local = dt_util.now()
+    utc_time = now.replace(hour=utc_hour, minute=0, second=0, microsecond=0)
+    local_hour = dt_util.as_local(utc_time)
+    return local_hour.replace(
+        year=now_local.year, month=now_local.month, day=now_local.day
+    )
```

Comment:
I think this can be simplified
```suggestion
    utc_now = dt_util.utcnow()
    utc_time = utc_now.replace(hour=utc_hour, minute=0, second=0, microsecond=0)
    return dt_util.as_local(utc_time)
```